### PR TITLE
Allow csv file uploads

### PR DESCRIPTION
**Pull request by tleraj**
The purpose of this PR is to close https://github.com/aqualinkorg/aqualink-app/issues/683.
We introduce the following new metrics (a migration for this is coming):
```typescript
PRESSURE = 'pressure',
PRECIPITATION = 'precipitation',
RH = 'rh',
WIND_GUST_SPEED = 'wind_gust_speed',
AIR_TEMPERATURE = 'air_temperature'
```
We use the same parsing function for csv files from `node-xlsx`, as we did for xlsx files. The only difference is on the timestamp generation, since meteorological data files have a different column format: `Date Time GMT +XX:XX`
We also introduce a stronger mimetype typing, `Mimetype`, so that we only accept specific file types.
In addition, we tweak the header row finding util function, refactored some already existing utils and created some others.

### TODOs
 - [x] Add metrics enum update migration
 - [x] Frontend work to allow csv files selection
 - [ ] ~~Investigate on using regexps to recognize sheet columns~~ Will be investigated in a future PR

### CODE REVIEW

**Comment by ericboucher** [comment]
@tleraj  we should also introduce a new type of temperature: AIR_TEMPERATURE

Indeed, met temp data should not be mixed up with water temp data

File: packages/api/scripts/upload-sonde-data.ts [hunk 1]
```diff
@@ -50,23 +51,28 @@ async function run() {
...
   const connection = await createConnection(config);
 
+  const isCsv = filePath.endsWith('.csv');
+
   logger.log('Uploading sonde data');
-  await uploadSondeData(
+  await uploadTimeSeriesData(
     filePath,
     last(filePath.split('/')),
+    isCsv
```
Comment (ericboucher) [inline packages/api/scripts/upload-sonde-data.ts:70]: I think this logic should happen within uploadTimeSeriesData

File: packages/api/src/utils/uploads/upload-sheet-data.ts [hunk 2]
```diff
@@ -71,6 +90,24 @@ export const fileFilter: MulterOptions['fileFilter'] = (
...
 };
 
+export const requiredHeadersForMimetype = (mimetype: Mimetype): string[] => {
+  switch (mimetype) {
+    case 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet':
+      return ['Date (MM/DD/YYYY)', 'Time (HH:mm:ss)', 'Time (Fract. Sec)'];
+    case 'text/csv':
+      return ['Date Time'];
+    default:
+      return [];
```
Comment (ericboucher) [inline packages/api/src/utils/uploads/upload-sheet-data.ts:100]: These are not linked to the mimetype. This is just the case with the data we are getting at the moment, but certainly not linked to the mimetype as we could have a csv or an xlsx with other requirements.

File: packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx [hunk 3]
```diff
@@ -27,6 +27,7 @@ type EnhancedSelectOption = SelectOption & { disabled?: boolean };
 
 const SENSOR_TYPES: EnhancedSelectOption[] = [
   { id: 0, name: "sonde", label: "Sonde data" },
+  { id: 1, name: "metlog", label: "Meteorological data" },
   // TODO: Enable these options once the API can support various sensors data upload.
```
Comment (ericboucher) [inline packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx:31]: can we remove this todo?

File: packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx [hunk 4]
```diff
@@ -27,6 +27,7 @@ type EnhancedSelectOption = SelectOption & { disabled?: boolean };
 
 const SENSOR_TYPES: EnhancedSelectOption[] = [
   { id: 0, name: "sonde", label: "Sonde data" },
+  { id: 1, name: "metlog", label: "Meteorological data" },
```
Comment (ericboucher) [inline packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx:30]: is it normal to have overlapping ids (1)?

File: packages/website/src/constants/metlogConfig.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  | "top_temperature"
+  | "pressure"
+  | "precipitation"
+  | "rh"
+  | "wind_gust_speed"
+>;
+
+export const metlogConfig: Record<MetlogMetricsKeys, MetlogConfig> = {
+  top_temperature: {
+    title: "Top Temperature",
```
Comment (ericboucher) [inline packages/website/src/constants/metlogConfig.ts:25]: This should be "Air Temperature"

File: packages/website/src/constants/metlogConfig.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  visibility: string;
+  order: number;
+  convert?: number;
+}
+
+export type MetlogMetricsKeys = Extract<
+  MetricsKeys,
+  | "wind_direction"
+  | "wind_speed"
+  | "top_temperature"
```
Comment (ericboucher) [inline packages/website/src/constants/metlogConfig.ts:16]: air_temperature

File: packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx (hunk 4 above)
Comment (tleraj) [reply packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx:30]: No, my bad :stuck_out_tongue:

File: packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx (hunk 3 above)
Comment (tleraj) [reply packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx:31]: It's just there to remind us that we should enable these options once our API supports it. But I can remove it :slightly_smiling_face:

File: packages/website/src/routes/SiteRoutes/UploadData/DropZone.tsx [hunk 7]
```diff
@@ -11,8 +11,8 @@ const GREY_COLOR = grey[500];
 const MAX_SIZE_MB = 10;
 const ACCEPTED_TYPES = [
   // TODO: Add these once our API supports them.
-  // { extension: "csv", mimeType: "text/csv" },
   // { extension: "xls", mimeType: "application/vnd.ms-excel" },
```
[the remark this answers is not in the document]
Comment (tleraj) [reply packages/website/src/routes/SiteRoutes/UploadData/DropZone.tsx:14]: Do we have any sample files so that I can test it?

File: packages/api/src/utils/uploads/upload-sheet-data.ts (hunk 2 above)
Comment (tleraj) [reply packages/api/src/utils/uploads/upload-sheet-data.ts]: I removed the mimetype dependency in https://github.com/aqualinkorg/aqualink-app/pull/685/commits/80145d833927c57a31d5dca4256456b63f8c3d86. Let me know what you think :slightly_smiling_face:

File: packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx (hunk 3 above)
Comment (ericboucher) [reply packages/website/src/routes/SiteRoutes/UploadData/Selectors.tsx]: Maybe rephrase it then?

File: packages/api/src/utils/uploads/upload-sheet-data.ts [hunk 8]
```diff
@@ -43,22 +43,51 @@ const metricsMapping: Record<string, Metric> = {
   'Temp °C': Metric.BOTTOM_TEMPERATURE,
   'Battery V': Metric.SONDE_BATTERY_VOLTAGE,
   'Cable Pwr V': Metric.SONDE_CABLE_POWER_VOLTAGE,
+  'Pressure, mbar (LGR S/N: 21067056, SEN S/N: 21062973)': Metric.PRESSURE,
```
Comment (ericboucher) [inline packages/api/src/utils/uploads/upload-sheet-data.ts:46]: So one of the reasons we need to use regex is because in the next files, the S/N will probably be different. So we need a better mapping logic that does not depend on `(LGR S/N: 21067056, SEN S/N: 21062973)`

**Comment by tleraj** [comment]
Currently, there is a bug when uploading meteorological data, like the ones in the `metlog240921_021021.csv`. With https://github.com/aqualinkorg/aqualink-app/pull/685/commits/1fec824df9db332fc0c85642325b0c2a6ef6c348, we do not look for exact matches with the `metricsMapping` dict, but we use a `csvHeader.toLowerCase().startsWith(metricKey.toLowerCase())` logic, so the following `renameKeys` util in https://github.com/aqualinkorg/aqualink-app/blob/1fec824df9db332fc0c85642325b0c2a6ef6c348/packages/api/src/utils/uploads/upload-sheet-data.ts#L116-L125

does not rename the keys properly.

**Comment by ericboucher** [comment]
@tleraj maybe move the renaming at the very beginning and with the new mapping logic. So that after you can simply use the final metrics names to do all the other checks. What do you think?

File: packages/api/src/utils/uploads/upload-sheet-data.ts [hunk 9]
```diff
@@ -43,22 +43,53 @@ const metricsMapping: Record<string, Metric> = {
   'Temp °C': Metric.BOTTOM_TEMPERATURE,
   'Battery V': Metric.SONDE_BATTERY_VOLTAGE,
   'Cable Pwr V': Metric.SONDE_CABLE_POWER_VOLTAGE,
+  'Pressure, mbar': Metric.PRESSURE,
+  'Rain, mm': Metric.PRECIPITATION,
+  'Temp, °C': Metric.AIR_TEMPERATURE,
```
Comment (ericboucher) [inline packages/api/src/utils/uploads/upload-sheet-data.ts:48]: You have two keys with the exact same name, maybe the mapping needs to be by source?

Comment (ericboucher) [reply packages/api/src/utils/uploads/upload-sheet-data.ts:48]: @tleraj

Comment (tleraj) [reply packages/api/src/utils/uploads/upload-sheet-data.ts]: Fixed in https://github.com/aqualinkorg/aqualink-app/pull/685/commits/a7021651e8987a7c15bfb81b12f5e1c20a8348d8

**Review by ericboucher — APPROVED** [review]
